Component esp32_touch not working in ESP32-S2

### The problem

As you can see in the logs below. Cannot compile correctly. In ESP32 it works fine. In ESP32-S2 it does not work. In the ESP-IDF framework it also does not work, not only in Arduino.
By the way, I found another bug. Not all touch pins from ESP32-S2 can be used for esp_touch. Please see the attachment. Only those that are in ESP32 can be used.
![screen](https://user-images.githubusercontent.com/10649327/219706314-fec5c37d-43ff-4815-90b7-f47ad9ec6366.jpg)


### Which version of ESPHome has the issue?

v2022.12.8

### What type of installation are you using?

Docker

### Which version of Home Assistant has the issue?

_No response_

### What platform are you using?

ESP32

### Board

Wemos ESP32 S2 Mini

### Component causing the issue

esp32_touch

### Example YAML snippet

```yaml
wifi:
  ssid: !secret wifi_iot_ssid
  password: !secret wifi_iot_password
  manual_ip:
    static_ip: 192.168.3.33
    gateway: 192.168.3.49
    subnet: 255.255.255.0

api:
  encryption:
    key: !secret api_encryption_key

ota:
  password: !secret esphome_api_password
    
esphome:
  name: test
  platformio_options:
    board_build.extra_flags:
      - "-DARDUINO_USB_CDC_ON_BOOT=0"
        
esp32:
  board: lolin_s2_mini
  framework:
    type: arduino

esp32_touch:
  setup_mode: true

binary_sensor: 
- platform: esp32_touch
  name: "PIN1"
  pin: GPIO12
  threshold: 1000
```


### Anything in the logs that might be useful for us?

```txt
INFO Reading configuration /config/test.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing test (board: lolin_s2_mini; framework: arduino; platform: platformio/espressif32 @ 5.2.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32S2 240MHz, 320KB RAM, 4MB Flash
 - toolchain-riscv32-esp @ 8.4.0+2021r2-patch3 
 - toolchain-xtensa-esp32s2 @ 8.4.0+2021r2-patch3
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
Dependency Graph
|-- WiFi @ 2.0.0
|-- ESPmDNS @ 2.0.0
|-- Update @ 2.0.0
|-- noise-c @ 0.1.4
|   |-- libsodium @ 1.10018.1
Compiling .pioenvs/test/src/esphome/components/esp32_touch/esp32_touch.cpp.o
Compiling .pioenvs/test/src/esphome/components/mdns/mdns_esp8266.cpp.o
Compiling .pioenvs/test/src/esphome/components/mdns/mdns_esp_idf.cpp.o
Compiling .pioenvs/test/src/esphome/components/mdns/mdns_rp2040.cpp.o
Compiling .pioenvs/test/src/esphome/components/network/util.cpp.o
Compiling .pioenvs/test/src/esphome/components/ota/ota_backend_arduino_esp32.cpp.o
Compiling .pioenvs/test/src/esphome/components/ota/ota_backend_arduino_esp8266.cpp.o
Compiling .pioenvs/test/src/esphome/components/ota/ota_backend_arduino_rp2040.cpp.o
Compiling .pioenvs/test/src/esphome/components/ota/ota_backend_esp_idf.cpp.o
Compiling .pioenvs/test/src/esphome/components/ota/ota_component.cpp.o
src/esphome/components/esp32_touch/esp32_touch.cpp: In member function 'virtual void esphome::esp32_touch::ESP32TouchComponent::setup()':
src/esphome/components/esp32_touch/esp32_touch.cpp:18:5: error: 'touch_pad_filter_start' was not declared in this scope
     touch_pad_filter_start(this->iir_filter_);
     ^~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/esp32_touch/esp32_touch.cpp:18:5: note: suggested alternative: 'touch_pad_filter_enable'
     touch_pad_filter_start(this->iir_filter_);
     ^~~~~~~~~~~~~~~~~~~~~~
     touch_pad_filter_enable
src/esphome/components/esp32_touch/esp32_touch.cpp:26:47: error: too many arguments to function 'esp_err_t touch_pad_config(touch_pad_t)'
     touch_pad_config(child->get_touch_pad(), 0);
                                               ^
In file included from src/esphome/components/esp32_touch/esp32_touch.h:12,
                 from src/esphome/components/esp32_touch/esp32_touch.cpp:3:
/config/.esphome/platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s2/include/driver/esp32s2/include/driver/touch_sensor.h:156:11: note: declared here
 esp_err_t touch_pad_config(touch_pad_t touch_num);
           ^~~~~~~~~~~~~~~~
src/esphome/components/esp32_touch/esp32_touch.cpp: In member function 'virtual void esphome::esp32_touch::ESP32TouchComponent::loop()':
src/esphome/components/esp32_touch/esp32_touch.cpp:117:7: error: 'touch_pad_read_filtered' was not declared in this scope
       touch_pad_read_filtered(child->get_touch_pad(), &value);
       ^~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/esp32_touch/esp32_touch.cpp:117:7: note: suggested alternative: 'touch_pad_read_raw_data'
       touch_pad_read_filtered(child->get_touch_pad(), &value);
       ^~~~~~~~~~~~~~~~~~~~~~~
       touch_pad_read_raw_data
src/esphome/components/esp32_touch/esp32_touch.cpp:119:7: error: 'touch_pad_read' was not declared in this scope
       touch_pad_read(child->get_touch_pad(), &value);
       ^~~~~~~~~~~~~~
src/esphome/components/esp32_touch/esp32_touch.cpp:119:7: note: suggested alternative: 'touch_pad_reset'
       touch_pad_read(child->get_touch_pad(), &value);
       ^~~~~~~~~~~~~~
       touch_pad_reset
src/esphome/components/esp32_touch/esp32_touch.cpp: In member function 'virtual void esphome::esp32_touch::ESP32TouchComponent::on_shutdown()':
src/esphome/components/esp32_touch/esp32_touch.cpp:142:5: error: 'touch_pad_filter_stop' was not declared in this scope
     touch_pad_filter_stop();
     ^~~~~~~~~~~~~~~~~~~~~
src/esphome/components/esp32_touch/esp32_touch.cpp:142:5: note: suggested alternative: 'touch_pad_fsm_stop'
     touch_pad_filter_stop();
     ^~~~~~~~~~~~~~~~~~~~~
     touch_pad_fsm_stop
src/esphome/components/esp32_touch/esp32_touch.cpp:143:5: error: 'touch_pad_filter_delete' was not declared in this scope
     touch_pad_filter_delete();
     ^~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/esp32_touch/esp32_touch.cpp:143:5: note: suggested alternative: 'touch_pad_filter_disable'
     touch_pad_filter_delete();
     ^~~~~~~~~~~~~~~~~~~~~~~
     touch_pad_filter_disable
src/esphome/components/esp32_touch/esp32_touch.cpp:155:77: error: too many arguments to function 'esp_err_t touch_pad_config(touch_pad_t)'
       touch_pad_config(child->get_touch_pad(), child->get_wakeup_threshold());
                                                                             ^
In file included from src/esphome/components/esp32_touch/esp32_touch.h:12,
                 from src/esphome/components/esp32_touch/esp32_touch.cpp:3:
/config/.esphome/platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s2/include/driver/esp32s2/include/driver/touch_sensor.h:156:11: note: declared here
 esp_err_t touch_pad_config(touch_pad_t touch_num);
           ^~~~~~~~~~~~~~~~
*** [.pioenvs/test/src/esphome/components/esp32_touch/esp32_touch.cpp.o] Error 1
========================== [FAILED] Took 4.62 seconds ==========================
```


### Additional information

_No response_

--- a/esphome/components/esp32_touch/binary_sensor.py
+++ b/esphome/components/esp32_touch/binary_sensor.py
@@ -76,9 +76,10 @@ def validate_touch_pad(value):
     if variant not in TOUCH_PADS:
         raise cv.Invalid(f"ESP32 variant {variant} does not support touch pads.")
 
-    if value not in TOUCH_PADS[variant]:
+    pads = TOUCH_PADS[variant]
+    if value not in pads:
         raise cv.Invalid(f"Pin {value} does not support touch pads.")
-    return value
+    return cv.enum(pads)(value)
 
 
 ESP32TouchBinarySensor = esp32_touch_ns.class_(
@@ -99,7 +100,7 @@ async def to_code(config):
     hub = await cg.get_variable(config[CONF_ESP32_TOUCH_ID])
     var = cg.new_Pvariable(
         config[CONF_ID],
-        TOUCH_PADS[config[CONF_PIN]],
+        config[CONF_PIN],
         config[CONF_THRESHOLD],
         config[CONF_WAKEUP_THRESHOLD],
     )

--- a/esphome/components/esp32_touch/binary_sensor.py
+++ b/esphome/components/esp32_touch/binary_sensor.py
@@ -1,5 +1,6 @@
 import esphome.codegen as cg
 import esphome.config_validation as cv
+from esphome.core import CORE
 from esphome.components import binary_sensor
 from esphome.const import (
     CONF_PIN,
@@ -7,6 +8,13 @@ from esphome.const import (
     CONF_ID,
 )
 from esphome.components.esp32 import gpio
+from esphome.components.esp32.const import (
+    KEY_ESP32,
+    KEY_VARIANT,
+    VARIANT_ESP32,
+    VARIANT_ESP32S2,
+    VARIANT_ESP32S3,
+)
 from . import esp32_touch_ns, ESP32TouchComponent
 
 DEPENDENCIES = ["esp32_touch", "esp32"]
@@ -15,22 +23,60 @@ CONF_ESP32_TOUCH_ID = "esp32_touch_id"
 CONF_WAKEUP_THRESHOLD = "wakeup_threshold"
 
 TOUCH_PADS = {
-    4: cg.global_ns.TOUCH_PAD_NUM0,
-    0: cg.global_ns.TOUCH_PAD_NUM1,
-    2: cg.global_ns.TOUCH_PAD_NUM2,
-    15: cg.global_ns.TOUCH_PAD_NUM3,
-    13: cg.global_ns.TOUCH_PAD_NUM4,
-    12: cg.global_ns.TOUCH_PAD_NUM5,
-    14: cg.global_ns.TOUCH_PAD_NUM6,
-    27: cg.global_ns.TOUCH_PAD_NUM7,
-    33: cg.global_ns.TOUCH_PAD_NUM8,
-    32: cg.global_ns.TOUCH_PAD_NUM9,
+    VARIANT_ESP32: {
+        4: cg.global_ns.TOUCH_PAD_NUM0,
+        0: cg.global_ns.TOUCH_PAD_NUM1,
+        2: cg.global_ns.TOUCH_PAD_NUM2,
+        15: cg.global_ns.TOUCH_PAD_NUM3,
+        13: cg.global_ns.TOUCH_PAD_NUM4,
+        12: cg.global_ns.TOUCH_PAD_NUM5,
+        14: cg.global_ns.TOUCH_PAD_NUM6,
+        27: cg.global_ns.TOUCH_PAD_NUM7,
+        33: cg.global_ns.TOUCH_PAD_NUM8,
+        32: cg.global_ns.TOUCH_PAD_NUM9,
+    },
+    VARIANT_ESP32S2: {
+        1: cg.global_ns.TOUCH_PAD_NUM1,
+        2: cg.global_ns.TOUCH_PAD_NUM2,
+        3: cg.global_ns.TOUCH_PAD_NUM3,
+        4: cg.global_ns.TOUCH_PAD_NUM4,
+        5: cg.global_ns.TOUCH_PAD_NUM5,
+        6: cg.global_ns.TOUCH_PAD_NUM6,
+        7: cg.global_ns.TOUCH_PAD_NUM7,
+        8: cg.global_ns.TOUCH_PAD_NUM8,
+        9: cg.global_ns.TOUCH_PAD_NUM9,
+        10: cg.global_ns.TOUCH_PAD_NUM10,
+        11: cg.global_ns.TOUCH_PAD_NUM11,
+        12: cg.global_ns.TOUCH_PAD_NUM12,
+        13: cg.global_ns.TOUCH_PAD_NUM13,
+        14: cg.global_ns.TOUCH_PAD_NUM14,
+    },
+    VARIANT_ESP32S3: {
+        1: cg.global_ns.TOUCH_PAD_NUM1,
+        2: cg.global_ns.TOUCH_PAD_NUM2,
+        3: cg.global_ns.TOUCH_PAD_NUM3,
+        4: cg.global_ns.TOUCH_PAD_NUM4,
+        5: cg.global_ns.TOUCH_PAD_NUM5,
+        6: cg.global_ns.TOUCH_PAD_NUM6,
+        7: cg.global_ns.TOUCH_PAD_NUM7,
+        8: cg.global_ns.TOUCH_PAD_NUM8,
+        9: cg.global_ns.TOUCH_PAD_NUM9,
+        10: cg.global_ns.TOUCH_PAD_NUM10,
+        11: cg.global_ns.TOUCH_PAD_NUM11,
+        12: cg.global_ns.TOUCH_PAD_NUM12,
+        13: cg.global_ns.TOUCH_PAD_NUM13,
+        14: cg.global_ns.TOUCH_PAD_NUM14,
+    },
 }
 
 
 def validate_touch_pad(value):
     value = gpio.validate_gpio_pin(value)
-    if value not in TOUCH_PADS:
+    variant = CORE.data[KEY_ESP32][KEY_VARIANT]
+    if variant not in TOUCH_PADS:
+        raise cv.Invalid(f"ESP32 variant {variant} does not support touch pads.")
+
+    if value not in TOUCH_PADS[variant]:
         raise cv.Invalid(f"Pin {value} does not support touch pads.")
     return value
 

